### PR TITLE
Move go-licenses to generate and separate generate into a frontend and backend component

### PR DESCRIPTION
**Pull request by zeripath**
The `go-licenses` make task introduced in #21034 is being run on make vendor
and occasionally causes an empty go-licenses file if the vendors need to
change. This should be moved to the generate task as it is a generated file.

Now because of this change we also need to split generation into two separate 
steps:

1. `generate-backend`
2. `generate-frontend`

In the future it would probably be useful to make `generate-swagger` part of `generate-frontend` but it's not tolerated with our .drone.yml

Ref #21034

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Review by jolheiser — APPROVED** [review]

**Review by delvh — APPROVED** [review]

**Comment by zeripath** [comment]
ah this will require a little more finesse...

File: modules/charset/invisible/generate.go [hunk 1]
```diff
@@ -63,6 +63,17 @@ func runTemplate(t *template.Template, filename string, data interface{}) error
 		verbosef("Bad source:\n%s", buf.String())
 		return fmt.Errorf("unable to format source: %w", err)
 	}
+
+	old, err := os.ReadFile(filename)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("failed to read old file %s because %w", filename, err)
+	} else if err == nil {
+		if res := bytes.Compare(bs, old); res == 0 {
```
Comment (delvh) [inline modules/charset/invisible/generate.go]:
```suggestion
		if bytes.Compare(bs, old) == 0 {
```

File: modules/charset/ambiguous/generate.go [hunk 2]
```diff
@@ -110,6 +110,17 @@ func runTemplate(t *template.Template, filename string, data interface{}) error
 		verbosef("Bad source:\n%s", buf.String())
 		return fmt.Errorf("unable to format source: %w", err)
 	}
+
+	old, err := os.ReadFile(filename)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("failed to read old file %s because %w", filename, err)
+	} else if err == nil {
+		if res := bytes.Compare(bs, old); res == 0 {
```
Comment (delvh) [inline modules/charset/ambiguous/generate.go]:
```suggestion
		if bytes.Compare(bs, old) == 0 {
```

File: Makefile [hunk 3]
```diff
@@ -278,7 +283,9 @@ TAGS_PREREQ := $(TAGS_EVIDENCE)
 endif
 
 .PHONY: generate-swagger
-generate-swagger:
+generate-swagger: $(SWAGGER_SPEC)
+
+$(SWAGGER_SPEC): $(GO_SOURCES_NO_BINDATA)
```
Comment (delvh) [inline Makefile:288]: Why not directly
```suggestion
generate-swagger: $(GO_SOURCES_NO_BINDATA)
```
?

Comment (zeripath) [reply Makefile:288]: Nope.

That won't do the right thing.

Try it

Comment (delvh) [reply Makefile:288]: Okay, I don't have to understand why, but I accept it.

**Review by delvh — APPROVED** [review]

File: Makefile (hunk 3 above)
Comment (zeripath) [reply Makefile:288]: It's to do with the dependencies - the SWAGGER_SPEC line is creating a file that acts as evidence for the build job. If all sources are older than the SWAGGER_SPEC file it doesn't need to be generated

File: modules/charset/ambiguous/generate.go [hunk 4]
```diff
@@ -110,6 +110,17 @@ func runTemplate(t *template.Template, filename string, data interface{}) error
 		verbosef("Bad source:\n%s", buf.String())
 		return fmt.Errorf("unable to format source: %w", err)
 	}
+
+	old, err := os.ReadFile(filename)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("failed to read old file %s because %w", filename, err)
+	} else if err == nil {
+		if bytes.Compare(bs, old) == 0 {
```
Comment (zeripath) [inline modules/charset/ambiguous/generate.go]:
```suggestion
		if bytes.Equal(bs, old) {
```

File: modules/charset/invisible/generate.go [hunk 5]
```diff
@@ -63,6 +63,17 @@ func runTemplate(t *template.Template, filename string, data interface{}) error
 		verbosef("Bad source:\n%s", buf.String())
 		return fmt.Errorf("unable to format source: %w", err)
 	}
+
+	old, err := os.ReadFile(filename)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("failed to read old file %s because %w", filename, err)
+	} else if err == nil {
+		if bytes.Compare(bs, old) == 0 {
```
Comment (zeripath) [inline modules/charset/invisible/generate.go]:
```suggestion
		if bytes.Equal(bs, old) {
```

File: Makefile [hunk 6]
```diff
@@ -702,16 +708,25 @@ install: $(wildcard *.go)
...
-backend: go-check generate $(EXECUTABLE)
+backend: go-check generate-backend $(EXECUTABLE)
 
+# We generate the backend before the frontend in case we in future we want to generate things in the frontend from generated files in backend
 .PHONY: generate
-generate: $(TAGS_PREREQ)
+generate: generate-backend generate-frontend
+
+.PHONY: generate-backend
+generate-backend: $(TAGS_PREREQ) generate-go generate-swagger
```
Comment (zeripath) [inline Makefile]:
```suggestion
generate-backend: $(TAGS_PREREQ) generate-go
```

unfortunately looks like we won't be able to fold the swagger in here at the moment as the checksums on the proxy don't match

**Comment by lunny** [comment]
make L-G-T-M work

**Comment by zeripath** [comment]
There's still an intermittent issue with the go-licenses make target that appears to drop the code.gitea.io license from time to time. 

@silverwind do you have any ideas what's happening here?

**Comment by silverwind** [comment]
Not sure I understand why these changes here were all necessary, but I'd probably just have dropped the `tidy` dependency of `vendor`. Reason being that `tidy` is unnecessary in the release pipeline as the `complicance` pipeline already ensures that `tidy` is consistent.

**Comment by silverwind** [comment]
> @silverwind do you have any ideas what's happening here?

Hmm, I could only imagine that the `go-licenses` tools fails to download sometimes, but I think it's using GOPROXY so it should work even if code.gitea.io is down.

**Comment by silverwind** [comment]
If automating this proves too difficult, we could also opt to make the target standalone without any dependants and just have it as a manual task after dependency update.

**Pull request merged**